DynamicActiveQuery.php

```diff
@@ -27,6 +28,44 @@ class DynamicActiveQuery extends ActiveQuery
      */
     private $dynamicColumn;
 
+    private $wrapped = false;
+
+    /**
+     * Convert index value to closurem that will get decoded dynamic attribute, in case if indexing attribute is dynamic
+     * @param callable|string $column
+     * @return $this
+     */
+    public function indexBy($column)
+    {
+        if ($this->asArray) {
+            /** @var \yii\db\ActiveRecord $modelClass */
+            $modelClass = $this->modelClass;
+            $this->indexBy = function ($row) use ($column, $modelClass) {
+                if (isset($row[$column])) {
+                    return $row[$column];
+                } elseif (method_exists($modelClass, 'dynamicColumn')) {
+                    $dynamicColumn = $modelClass::dynamicColumn();
+                    if (isset($row[$dynamicColumn])) {
+                        $dynamicAttributes = DynamicActiveRecord::dynColDecode($row[$dynamicColumn]);
+
+                        if ($value = $this->getDotNotatedValue($dynamicAttributes, $column)) {
```

Comment:
please don't put assignments in if conditions. with that rule we can use a simple linter and eliminate a whole class of mistake